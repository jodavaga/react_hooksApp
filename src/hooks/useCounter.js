import { useState } from "react";

export const useCounter = ( initialState = 10) => {

    const [state, setState] = useState(initialState);

    const increment = ( num = 1) => {
        setState( state + num );
    }

    const decrement = ( num = 1 ) => {
        setState( state - num );
    }

    const reset = () => {
        setState( initialState );
    } 

    return {
        state,
        increment,
        decrement,
        reset
    }
}