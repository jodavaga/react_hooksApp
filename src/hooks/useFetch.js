import { useState, useEffect, useRef } from 'react';

export const useFetch = ( url ) => {

    const [ state, setState ] = useState({ data: null, loading: true, error: null });

    // useRef to manage mount-unmount
    const isMounted = useRef(true);

    useEffect(() => {

        return () => {
            console.log("unmounted");
            isMounted.current = false;
        }
    }, [])

    if (!url) {
        throw new Error('Url required!');
    }

    useEffect(() => {

        setState({
            ...state,
            loading: true
        })
        
        fetch( url )
            .then( resp => resp.json() )
            .then( data => {

                // Only trigger setState if is mounted.
                if ( isMounted.current ) {
                    setState({
                        data,
                        loading: false,
                        errors: null
                    })
                } else {
                    console.log('setState not triggered')
                }
            })
            .catch( error => {
                setState({
                    data: null,
                    loading: false,
                    error
                })
            })
    }, [ url ]);

    return state;

}