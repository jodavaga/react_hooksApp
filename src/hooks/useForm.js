import { useState } from 'react';


// Manage forms using target names as a input form ref and it's values.

export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);

    const handleInputChange = ({ target }) => {
        setValues({
            ...values,
            [ target.name ]: target.value 
        })
    }

    const reset = () => {
        setValues( initialState );
    }

    return [ values, handleInputChange, reset ];
}