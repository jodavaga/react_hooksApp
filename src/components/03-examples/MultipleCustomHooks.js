import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

    const { counter, increment } = useCounter( 1 );
    const state = useFetch( `https://www.breakingbadapi.com/api/quotes/${ counter }` );


    // console.log(state);

    const { data, loading, error } = state;

    // destructured data object. Evaluate, if not null, return data[0]
    const { author, quote } = !!data && data[0];

    return (
        <div className="container">
            <h1>Breaking Bad Quotes</h1>
            <hr />
            {
                error && <div className="alert alert-danger"> Algo salio mal </div>
            }

            { loading 
                ? 
                    (
                        <div className="alert alert-info">Loading...</div>
                    )  
                :   
                    (
                        <div className="card">
                            <div className="card-body">
                                <blockquote className="blockquote text-right">
                                    <p>{ quote }</p>
                                    <footer className="blockquote-footer">{ author }</footer>
                                </blockquote>
                            </div>
                        </div>
                    )       
            }

            <button
                className="btn btn-primary mt-3"
                onClick={ () => increment() }
            >Next</button>

        </div>
    )
}