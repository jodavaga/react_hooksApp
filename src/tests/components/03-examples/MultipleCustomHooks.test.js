import React from 'react';
// just to get helps
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Test on <MultipleCustomHooks />', () => {
    
    useCounter.mockReturnValue({
        counter: 10,
        increment: () => {}
    });

    test('should render component initialState', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });
        

        const wrapper = shallow( <MultipleCustomHooks /> );
        expect( wrapper ).toMatchSnapshot();

    });

    test('should render fetch information', () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: 'Jose D',
                quote: 'Hola mundo!'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );

        // console.log( wrapper.html() );

        expect( wrapper.find('.blockquote').exists() ).toBe(true);
        expect( wrapper.find('p').text().trim() ).toBe('Hola mundo!');
        expect( wrapper.find('footer').text().trim() ).toBe('Jose D');
    });
    
    
})


