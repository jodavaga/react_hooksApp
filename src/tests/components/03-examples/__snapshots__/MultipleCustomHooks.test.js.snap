// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <MultipleCustomHooks /> should render component initialState 1`] = `
<div
  className="container"
>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary mt-3"
    onClick={[Function]}
  >
    Next
  </button>
</div>
`;
